### PR TITLE
Update minimum support Java version from 8 to 11

### PR DESCRIPTION
**Pull request by reta**
Signed-off-by: Andriy Redko <andriy.redko@aiven.io>

Fixes https://github.com/mockito/mockito/issues/2798

## Checklist

 - [X] Read the [contributing guide](https://github.com/mockito/mockito/blob/main/.github/CONTRIBUTING.md)
 - [X] PR should be motivated, i.e. what does it fix, why, and if relevant how
 - [ ] If possible / relevant include an example in the description, that could help all readers
       including project members to get a better picture of the change
 - [ ] Avoid other runtime dependencies
 - [X] Meaningful commit history ; intention is important please rebase your commit history so that each
       commit is meaningful and help the people that will explore a change in 2 years
 - [X] The pull request follows coding style
 - [X] Mention `Fixes #<issue number>` in the description _if relevant_
 - [X] At least one commit should mention `Fixes #<issue number>` _if relevant_

### CODE REVIEW

File: subprojects/errorprone/errorprone.gradle
Comment (reta) [inline subprojects/errorprone/errorprone.gradle:21]: Needed by `errorprone`

File: subprojects/errorprone/errorprone.gradle
Comment (reta) [inline subprojects/errorprone/errorprone.gradle:19]: `-Xbootclasspath/p:` is gone (https://openjdk.org/jeps/261)

**Comment by reta** [comment]
@TimvdLippe started the initial work here, thank you

**Comment by reta** [comment]
It seems like I run into https://github.com/mojohaus/animal-sniffer/issues/172 again (but this time using 1.22) :(

**Comment by TimvdLippe** [comment]
Thanks for the PR, this is great! We will need to publish `mockito-subclass` in Mockito 4 first (https://github.com/mockito/mockito/issues/2589#issuecomment-1328262373), so until then we can't merge this PR.

Seems like CI is unhappy about our Javadoc. I couldn't find the failing error, but we have to build our Javadoc with no errors.

**Comment by reta** [comment]
> Thanks for the PR, this is great! We will need to publish `mockito-subclass` in Mockito 4 first ([#2589 (comment)](https://github.com/mockito/mockito/issues/2589#issuecomment-1328262373)), so until then we can't merge this PR.
> 

:+1: 

> Seems like CI is unhappy about our Javadoc. I couldn't find the failing error, but we have to build our Javadoc with no errors.

Thanks @TimvdLippe , I will certainly take a look

**Comment by reta** [comment]
@TimvdLippe looks like regression https://github.com/mojohaus/animal-sniffer/issues/172 is back in the new version, I will try to figure that out.

```
[Undefined reference | android-api-level-24-7.0_r2] org.mockito.internal.creation.bytebuddy.(InlineBytecodeGenerator.java:1)
  >> java.lang.instrument.Instrumentation
> Task :animalsnifferMain FAILED

[Undefined reference | android-api-level-24-7.0_r2] org.mockito.internal.creation.bytebuddy.(InlineDelegateByteBuddyMockMaker.java:1)
  >> java.lang.instrument.Instrumentation

[Undefined reference | android-api-level-24-7.0_r2] org.mockito.internal.creation.proxy.(MethodHandleProxy.java:1)
  >> java.lang.invoke.MethodHandles.Lookup

[Undefined reference | android-api-level-24-7.0_r2] org.mockito.internal.creation.proxy.(MethodHandleProxy.java:1)
  >> java.lang.invoke.MethodHandle

[Undefined reference | android-api-level-24-7.0_r2] org.mockito.internal.util.reflection.(InstrumentationMemberAccessor.java:1)
  >> java.lang.instrument.Instrumentation

[Undefined reference | android-api-level-24-7.0_r2] org.mockito.internal.util.reflection.(InstrumentationMemberAccessor.java:1)
  >> java.lang.invoke.MethodHandles.Lookup

Note: Some input files use or override a deprecated API.

Note: Recompile with -Xlint:deprecation for details.
> Task :errorprone:compileTestJava
```

**Comment by reta** [comment]
> @TimvdLippe looks like regression [mojohaus/animal-sniffer#172](https://github.com/mojohaus/animal-sniffer/issues/172) is back in the new version, I will try to figure that out.

Huh ... found a workaround here: https://github.com/mockito/mockito/pull/2804/files#diff-49a96e7eea8a94af862798a45174e6ac43eb4f8b4bd40759b5da63ba31ec3ef7R109, not ideal but the only one which works

File: subprojects/kotlinTest/src/test/kotlin/org/mockito/kotlin/CircularityTest.kt
Comment (reta) [inline subprojects/kotlinTest/src/test/kotlin/org/mockito/kotlin/CircularityTest.kt:1]: Not related, but build complains about missed headers, adding them here

File: subprojects/errorprone/errorprone.gradle
Comment (reta) [inline subprojects/errorprone/errorprone.gradle:20]: Fixed `errorprone` tests (were only run with JDK8 before)

**Comment by reta** [comment]
image 

Should be good to go :)

File: src/main/java/org/mockito/internal/debugging/LocationFactory.java
Comment (TimvdLippe) [inline src/main/java/org/mockito/internal/debugging/LocationFactory.java:26]: I know you were initially working on this PR to get these changes in. However, I would like to include these changes in Mockito 5.1.0, so to avoid pulling in too many changes into 1 go. Can you split out these changes into a separate PR so that we can land that? We can then also consider cleaning up some more of these Java 8 utility classes such as `JavaEightUtil` (https://github.com/mockito/mockito/blob/e2e628918518b73de7381692f28556ac6caa4710/src/main/java/org/mockito/internal/stubbing/defaultanswers/ReturnsEmptyValues.java#L132)

**Comment by TimvdLippe** [comment]
All of these changes look good! I would like to land the functional changes in a separate PR and in Mockito 5.1.0 to minimize impact. Other than that, we are mostly blocked on `mockito-subclass` (https://github.com/mockito/mockito/issues/2589#issuecomment-1328262373) before we can merge this PR.

File: src/main/java/org/mockito/internal/debugging/LocationFactory.java
Comment (reta) [reply src/main/java/org/mockito/internal/debugging/LocationFactory.java:26]: @TimvdLippe sure, I can split that, thanks!

Comment (reta) [reply src/main/java/org/mockito/internal/debugging/LocationFactory.java]: @TimvdLippe I reverted the changes, would you be open to have a plugin for `LocationFactory` so the alternative implementation could be supplied? (with revert, we are back to reflection <-> `SecurityManager` issues we've started with). I will open separate issue / pull request, should be a low risk change from implementation perspective, but I am wondering how it is aligned with the project.

File: src/main/java/org/mockito/internal/debugging/LocationImpl.java
Comment (TimvdLippe) [inline src/main/java/org/mockito/internal/debugging/LocationImpl.java:24]: This should probably have been removed from this PR as well, but I don't see any harm in including an unused class for now. Since I don't want another review cycle, let's merge this now so we can get things going! Thanks 😄

**Review by TimvdLippe — APPROVED** [review]
This is great work, thank you so much!

**Pull request merged**
